### ISSUE TITLE
LoRA adapter merging gives loss/output the similar to base (untrained) model

### ISSUE THREAD

**Issue by jalola** [NVIDIA-NeMo/Automodel#1226]
**Describe the bug**

I train Qwen3 model + LoRA using config similar to [this one](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/qwen/qwen3_moe_30b_lora.yaml) except that I use ChatDataset on my own dataset and with Qwen/Qwen3-30B-A3B-Instruct-2507 instead of Qwen/Qwen3-30B-A3B

- After the adapter was trained for 200 hundred steps (global batch size is 64), the train loss decreased from 1.6 to 0.22, val loss followed similar trend.
- After that, I merged the adapter (`LOWEST_VAL/model/`) with base model by this be low
- Then I got very bad result on test set (0.5 for micro f1) and I tested with base model without training (Qwen/Qwen3-30B-A3B-Instruct-2507), it results in similar micro f1. I suspect the the merging does not work or the saving adapter is malfunctioned that the adapter has no effect on final model. Or the LoRa alpha is not applied correctly (I used alpha: 32)?
- I also load the merged model and run over the validation dataset and compute the cross entropy loss, I got the loss of 2.7889 which is similar to the loss of untrained model with loss at 2.8697.

```
import gc
import os
from pathlib import Path

import fire
import torch
from peft import AutoPeftModelForCausalLM
from transformers import AutoTokenizer


def is_peft_model(model_path: str) -> bool:
    """Check if model is PEFT by looking for adapter_config.json."""
    adapter_config_path = Path(model_path) / "adapter_config.json"
    return adapter_config_path.exists()


def merge_and_save(
    adapter_path: str,
    output_path: str,
    device_map: str = "auto",
):
    """Merge PEFT adapter with base model and save the merged model.

    Args:
        adapter_path: Path to the PEFT adapter checkpoint
        output_path: Path to save the merged model
        device_map: Device map for model loading
    """

    # Check if it's a PEFT model
    if not is_peft_model(adapter_path):
        raise ValueError(
            f"Not a PEFT model: {adapter_path}\n"
            "adapter_config.json not found. This script only works with PEFT models."
        )

    print(f"\nLoading PEFT model from: {adapter_path}")
    model = AutoPeftModelForCausalLM.from_pretrained(
        adapter_path, torch_dtype="auto", device_map=device_map, trust_remote_code=True
    )

    print("Loading tokenizer...")
    tokenizer = AutoTokenizer.from_pretrained(adapter_path)

    print("\nMerging adapter with base model...")
    merged_model = model.merge_and_unload()

    print(f"\nSaving merged model to: {output_path}")
    os.makedirs(output_path, exist_ok=True)
    merged_model.save_pretrained(output_path)
    tokenizer.save_pretrained(output_path)
```

**Steps/Code to reproduce bug**

I use A100 8x80GB

- Train adapter with this [config](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/qwen/qwen3_moe_30b_lora.yaml)
- Merge adapter `LOWEST_VAL/model/` with base model
- Run over validation data to get the loss

**Expected behavior**

The trained and merged model should have better result than untrained model

**Additional context**

I didn't try to run FFT, hopefully it saves the correct trained model

I also see optim dir, is there any way to skip the optim saving? it takes some quite space

**Comment by akoumpa** [comment]
Hi @jalola , thank you for the detailed issue,

A few things from a first read:
- the instruct model has "think" tokens, that make tokenization harder than the usual base models, so that could be one source of error if not handled
- the MoE + PeFT path is still being validated. 

I will try to reproduce this tomorrow and come back with my findings.
